### PR TITLE
[Java] Add eager operation builder

### PR DESCRIPTION
**Pull request by karllessard**
PR #3 of 5 of the Java eager environment epic.

This one adds a new implementation of `OperationBuilder` for eager operations. It replicates pretty much the `GraphOperationBuilder` but native calls are redirected to the TFE C API endpoints instead.

CC: @sjamesr

### CODE REVIEW

**Comment by rthadur** [comment]
@angersson can you please help review this PR ?

**Comment by angersson** [comment]
@rthadur I think @sjamesr is a better fit -- I don't know how TF Java works.

**Comment by rthadur** [comment]
> @rthadur I think @sjamesr is a better fit -- I don't know how TF Java works.

Sure , thank you @angersson

File: tensorflow/java/src/main/java/org/tensorflow/EagerOperationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+package org.tensorflow;
+
+import java.nio.charset.Charset;
+
+/** 
+ * An {@link OperationBuilder} for building {@link Operation}s that are executed eagerly.
```
Comment (sjamesr) [inline tensorflow/java/src/main/java/org/tensorflow/EagerOperationBuilder.java]: I have a slight preference for `{@link Operation Operations}` over `{@link Operation}s` but maybe that's just me.

File: tensorflow/java/src/main/java/org/tensorflow/EagerOperationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+  }
+
+  @Override
+  public OperationBuilder setAttr(String name, String value) {
+    return setAttr(name, value.getBytes(Charset.forName("UTF-8")));
+  }
+
+  @Override
+  public OperationBuilder setAttr(String name, String[] values) {
+    Charset utf8 = Charset.forName("UTF-8");
```
Comment (sjamesr) [inline tensorflow/java/src/main/java/org/tensorflow/EagerOperationBuilder.java]: `StandardCharsets.UTF_8`?

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+  }
+  return rethandles;
+}
+
+JNIEXPORT void JNICALL Java_org_tensorflow_EagerOperationBuilder_setDevice(
+    JNIEnv* env, jclass clazz, jlong op_handle, jstring device_name) {
+  TFE_Op* op = requireOp(env, op_handle);
+  if (op == nullptr) return;
+  const char* cname = env->GetStringUTFChars(device_name, nullptr);
+  TF_Status* status = TF_NewStatus();
```
Comment (sjamesr) [inline tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: is this status leaked?

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+  TFE_OpSetDevice(op, cname, status);
+  env->ReleaseStringUTFChars(device_name, cname);
```
Comment (sjamesr) [inline tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: status return value not checked?

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+
+JNIEXPORT void JNICALL Java_org_tensorflow_EagerOperationBuilder_addInputList(
+    JNIEnv* env, jclass clazz, jlong op_handle, jlongArray input_handles) {
+  TFE_Op* op = requireOp(env, op_handle);
+  if (op == nullptr) return;
+  jlong* cinput_handles = env->GetLongArrayElements(input_handles, nullptr);
+  size_t num_inputs = static_cast<size_t>(env->GetArrayLength(input_handles));
+  TFE_TensorHandle* tensor_handles[num_inputs];
+  for (int i = 0; i < num_inputs; ++i) {
+    tensor_handles[i] = requireTensorHandle(env, cinput_handles[i]);
```
Comment (sjamesr) [inline tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: should we clean up and return if requireTensorHandle returns nullptr? Maybe this loop will call throwException multiple times. I'm not sure what the semantics are if we do that.

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+  std::unique_ptr<int[]> cnum_dims;
+  const int num_dims_length = env->GetArrayLength(num_dims);
+  if (num_dims_length > 0) {
+    const int shapes_length = env->GetArrayLength(shapes);
+    cshapes.reset(new int64_t[shapes_length]);
+    cdims.reset(new const int64_t*[num_dims_length]);
+    cnum_dims.reset(new int[num_dims_length]);
+    jlong* shapes_elems =
+        static_cast<jlong*>(env->GetPrimitiveArrayCritical(shapes, nullptr));
+    std::memcpy(cshapes.get(), shapes_elems, shapes_length << 3);
```
Comment (sjamesr) [inline tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: what's the deal with the left shift?

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+      cnum_dims[i] = static_cast<int>(num_dims_elems[i]);
+      cdims[i] = cshapes_ptr;
+      if (cnum_dims[i] > 0) {
+        cshapes_ptr += cnum_dims[i];
+      }
+    }
+    env->ReleasePrimitiveArrayCritical(num_dims, num_dims_elems, JNI_ABORT);
+  }
+  const char* cname = env->GetStringUTFChars(attr_name, nullptr);
+  TF_Status* status = TF_NewStatus();
```
Comment (sjamesr) [inline tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: status leaked?

File: tensorflow/java/src/main/java/org/tensorflow/EagerOperationBuilder.java (hunk 1 above)
Comment (karllessard) [reply tensorflow/java/src/main/java/org/tensorflow/EagerOperationBuilder.java]: Np, I'll change it

File: tensorflow/java/src/main/java/org/tensorflow/EagerOperationBuilder.java (hunk 2 above)
Comment (karllessard) [reply tensorflow/java/src/main/java/org/tensorflow/EagerOperationBuilder.java]: Yep, that would fit too. The "UTF-8" literal is used in many other places in the client though, but I suggest we update those later as we go.

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc (hunk 3 above)
Comment (karllessard) [reply tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: Good catch, fixing it.

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc (hunk 6 above)
Comment (karllessard) [reply tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: Ah, beautiful JNI code... that basically a optimized way to copy `shapes_length` 64-bits values from `shapes_elems` to `cshapes` (i << 3 == i * 8 bytes)

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc (hunk 7 above)
Comment (karllessard) [reply tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: good catch again, thanks!

File: tensorflow/java/src/main/native/eager_operation_builder_jni.cc (hunk 5 above)
Comment (karllessard) [reply tensorflow/java/src/main/native/eager_operation_builder_jni.cc]: I think it would be wiser to handle it, yes. I'll make the change.

**Comment by karllessard** [comment]
Thanks for the review @sjamesr , I just pushed my changes.

**Review by sjamesr — DISMISSED** [review]
Thank you!

**Comment by karllessard** [comment]
Ok @sjamesr, I replaced the dynamic-size array allocation by a smart pointer, looks like that's what is being used elsewhere in the JNI code. Strange that my local GCC did not complain about it...

**Review by sjamesr — DISMISSED** [review]

**Review by sjamesr — APPROVED** [review]

**Review by sjamesr — APPROVED** [review]

**Pull request merged**